lib/semian/lru_cache.rb

```diff
@@ -0,0 +1,51 @@
+class LRUCache
+  extend Forwardable
+  def_delegators :@table, :size, :count
+  def_delegators :@table, :empty?
+  def_delegators :@table, :keys
+  def_delegators :@table, :clear
+
+  attr_accessor :max_items, :table
+  def initialize(max_items)
+    @max_items = max_items
+    @table = {}
+  end
+
+  def set(key, value)
+    @table.delete(key)
+    @table[key] = value
+    clear_unused_resources
+    # @table.shift if @table.length > @max_items
+    value
+  end
+
+  def get(key)
+    if found = @table.delete(key)
+      @table[key] = found
+    else
+      nil
+    end
+  end
+
+  def []=(key, value)
+    set(key, value)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  def delete(key)
+    @table.delete(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    @table.each do |key, value|
+      next unless value.instance_of?(Semian::ProtectedResource)
+      @table.delete(key) unless value&.circuit_breaker.in_use? && !!value&.bulkhead
```

Comment:
This in `in_use?` pattern is perfect. Makes a ton of sense to me. What may be a tad bit surprising to some users is that circuit breaker resources are cleared, but those with bulkheads are _never_ cleared. I think it's reasonable to do, however, clearing a bulkhead should not be a big deal. That said, it's somewhat harder to reason about a newly created bulkhead that's "warming up" as the workers register, than an established one—so I think it makes sense to leave these out for now.

I think you should delegate that decision down to the `bulkhead` though. You can define a method in `ext/semian/semian.c` like so:

```
VALUE
semian_resource_in_use(VALUE self)
{
  return Qtrue;
}
```

And then attach it to the object at the bottom:

```
rb_define_method(cResource, "in_use?", semian_resource_in_use, 0);
```

However, it's not unreasonable in that function to call `semian_resource_tickets` and if any tickets are claimed, returning `false`. As mentioned earlier, the amount of churn would thus be concerning. `in_use?` could be `false` if _it's not currently claimed_ and _it hasn't been claimed in the past 60 seconds_. The latter information can be found through the [`semctl(2)`](http://man7.org/linux/man-pages/man2/semctl.2.html) system call, that's already used in the C extension.

Additionally, it would make sense to start LRUing the bulkheads if you reach the ceiling on the number of semaphores allowed on the system, a number you can obtain with the `IPC_INFO` call to `semctl(2)`, see again the [man page](http://man7.org/linux/man-pages/man2/semctl.2.html). Being close to this ceiling at claim time, could invoke this method.

To boil this down: You can do much more here, but, you're hitting diminishing very quickly. I'd suggest just putting `in_use?` to return `Qtrue` in the C extension for the time being.

Reply:
I think it would be good custom here to call `destroy` on all the resource (the circuit + bulkhead, even if the latter doesn't exist) before deleting it from the hash. I don't expect that to make any difference now, but maybe it would in the future if the Circuit starts doing things like modifying timeouts permanently and needs to be cleaned up 🤷‍♂️